providers/implementations/signature/slh_dsa_sig.c.in

```diff
@@ -39,6 +41,70 @@ static OSSL_FUNC_signature_dupctx_fn slh_dsa_dupctx;
 static OSSL_FUNC_signature_set_ctx_params_fn slh_dsa_set_ctx_params;
 static OSSL_FUNC_signature_settable_ctx_params_fn slh_dsa_settable_ctx_params;
 
+#ifdef FIPS_MODULE
+static int slh_dsa_sig_kat_pass = 0;
+#endif
+
+static int slh_dsa_self_check(OSSL_LIB_CTX *libctx)
+{
+    if (!ossl_prov_is_running())
+        return 0;
+
+#ifdef FIPS_MODULE
+    /*
+     * until this lock is initialized we do not want to recurse into
+     * the self testing infrastructure
+     */
+    if (!slh_dsa_sig_kat_pass) {
+        int state;
+
+        /*
+         * We consider the SHA2 variant covered by the keygen test
+         * NOTE: usually this is a noop as the same test will have been
+         * previously performed by the keygen code
+         */
+        state = FIPS_kat_on_demand(libctx, OSSL_ONDEMAND_KAT_ASYM_KEYGEN,
+                                   "SLH-DSA-SHA2-128f");
+        switch (state) {
+        case -1:
+            /*
+             * self test in progress for this thread so we let operations
+             * continue without changing any other state
```

Comment:
You cannot use the algorithm before the self test has finished, so wouldnt it have to wait here?